### ISSUE TITLE
Repeating Scheduled Maintenance Peiords

### ISSUE THREAD

**Issue by jessereynolds** [flapjack/flapjack#803]
Lets say you know your service is going to be down every night at 1am for up to an hour. Wouldn't it be great if this could be scheduled as a repeating maintenance period in Flapjack. 

We already use the iCalendar standard to define repeating time periods within Notification Rules for controlling when alerts are generated. It would make sense to allow iCalendar repeating time periods be used for scheduled maintenances as well. 

This would be a 2.x thing. 

Perhaps it doesn't fit as a scheduled maintenance due to existing semantics in the api around this object type, in which case it might need to be called something else, "regular downtime"? Hopefully not though.

**Comment by ali-graham** [comment]
I was thinking about this a little while back; it could make querying whether a check was currently in maintenance slower though, as you'd need to instantiate all schedule rules and see whether they applied any time you needed to know that. What might need to happen is to save out a normalised list of ranges of, e.g. time periods since the start of the day/week/month that the check would be in maintenance for, alongside the rule schedule definition.

**Comment by jessereynolds** [comment]
Yeah that makes sense @ali-graham ... the trigger for which would be a notification for a check, if the current period hasn't had its normalised set of ranges created then it does and does that before progressing? ... I suppose they could be regular scheduled maintenance periods (current, simple ones) that are instantiated?

**Comment by ali-graham** [comment]
Yeah, wherever the current check for `in_scheduled_maintenance?` is called. They'd have to be "virtual", relative ones (along with a means of determining the base point for the relativity) due to the possibility of indefinite lifetime of some of the schedules (e.g. one hour every 2 weeks from now until the end of time).

**Comment by jessereynolds** [comment]
Timezone (and daylight savings time) is an interesting consideration. With Notification Rules we store no timezone with the time restrictions, and interpret them according to the recipient's configured timezone (falling back to the default timezone in the flapjack configuration). What should we do with scheduled maintenances? Should they have a specific timezone, unlike time restrictions on notification rules?

**Comment by ali-graham** [comment]
I think it's reasonable to assume that maintenance times are in the server's timezone/DST state.

**Comment by jessereynolds** [comment]
> I think it's reasonable to assume that maintenance times are in the server's timezone/DST state.

That's fine until you have services in a remote datacentre, in a different timezone to Flapjack (with different DST dates), with a nightly batch job at 4am localtime. :-)

**Comment by jessereynolds** [comment]
But yes, we could decide to only support the timezone flapjack is running in (hopefully UTC).
